nil/internal/execution/block_generator.go

```diff
@@ -102,27 +95,48 @@ func (g *BlockGenerator) Rollback() {
 	g.rwTx.Rollback()
 }
 
-func (p *BlockGenerator) CollectGasPrices(prevBlockHash common.Hash, shardHashes []common.Hash) []types.Uint256 {
+func (p *BlockGenerator) CollectGasPrices(prevBlockId types.BlockNumber) []types.Uint256 {
 	if !p.params.ShardId.IsMainShard() {
 		return nil
 	}
 
+	// Basically we load configuration from block.MainChainHash.
+	// But for main shard this value should be block.PrevBlock.
+	// The first block uses configuration from itself.
+	configBlockId := prevBlockId
+	if prevBlockId != 0 {
+		configBlockId--
+	}
+
+	mainBlock, err := db.ReadBlockByNumber(p.rwTx, types.MainShardId, configBlockId)
```

Comment:
It is worth noting that we need to be more strict about not addressing things by "height", as in presence of forks these heights will resolve to something you may not expect. Not for this PR for sure, but just for future.

Reply:
This part can be replaced with two `ReadBlocks` - get current block by hash (lookup prevBlock by hash and then lookup its prevBlock by hash).